hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieSimpleIndex.java

```diff
@@ -0,0 +1,263 @@
+   */
+  @VisibleForTesting
+  List<Tuple2<String, String>> loadInvolvedFileIds(List<String> partitions, final JavaSparkContext jsc,
+                                                   final HoodieTable hoodieTable) {
+
+    // Obtain the latest data files from all the partitions.
+    List<Pair<String, String>> partitionPathFileIDList =
+        jsc.parallelize(partitions, Math.max(partitions.size(), 1)).flatMap(partitionPath -> {
+          Option<HoodieInstant> latestCommitTime =
+              hoodieTable.getMetaClient().getCommitsTimeline().filterCompletedInstants().lastInstant();
+          List<Pair<String, String>> filteredFiles = new ArrayList<>();
+          if (latestCommitTime.isPresent()) {
+            filteredFiles = hoodieTable.getROFileSystemView()
+                .getLatestDataFilesBeforeOrOn(partitionPath, latestCommitTime.get().getTimestamp())
+                .map(f -> Pair.of(partitionPath, f.getFileId())).collect(toList());
+          }
+          return filteredFiles.iterator();
+        }).collect();
+    return partitionPathFileIDList.stream()
+        .map(pf -> new Tuple2<>(pf.getKey(), pf.getValue())).collect(toList());
+  }
+
+  /**
+   * Find <HoodieKey, HoodieRecordLocation> for all incoming HoodieKeys
+   */
+  private JavaPairRDD<HoodieKey, HoodieRecordLocation> findMatchingFilesForRecordKeys(JavaSparkContext jsc,
+                                                                                      List<Tuple2<String, String>> partitionToFileIndexInfo,
+                                                                                      JavaPairRDD<String, String> partitionRecordKeyPairRDD, HoodieTable hoodieTable) {
+    // Step 1: Create JavaPairRDD< Tuple2<PartitionPath, RecordKey>, Optional<HoodieRecordLocation> > from input with  Optional<HoodieRecordLocation> as Empty.
+    JavaPairRDD<Tuple2<String, String>, Optional<HoodieRecordLocation>> partitionRecordPairs =
+        partitionRecordKeyPairRDD.mapToPair((PairFunction<Tuple2<String, String>, Tuple2<String, String>, Optional<HoodieRecordLocation>>) stringStringTuple2
+            -> new Tuple2(stringStringTuple2, Optional.absent()));
+
+    // Step 2: Create JavaRDD< Tuple2 <Partition, FileId> > from partitions to be touched
+    JavaRDD<Tuple2<String, String>> partitionFileIdTupleRDD = jsc.parallelize(partitionToFileIndexInfo);
+
+    // Step 3: For each partiion, fileId Tuple -> Fetch RDD of triplets ( Tuple2 <PartitionPath, recordKey>, HoodieRecordLocation )
+    JavaRDD<Tuple2<Tuple2<String, String>, Option<HoodieRecordLocation>>> partitionFileIdLocationEntries = partitionFileIdTupleRDD.flatMap(partitionFileIdTuple -> {
+      HoodieDataFile latestDataFile = getLatestDataFile(hoodieTable, Pair.of(partitionFileIdTuple._1, partitionFileIdTuple._2));
+      List<Pair<Pair<String, String>, Option<HoodieRecordLocation>>> resultList = ParquetUtils.fetchRecordKeyPartitionPathFromParquet(hoodieTable.getHadoopConf(), new Path(latestDataFile.getPath()),
```

Comment:
so this only works for COW ? or assumed inserts never go to logs? I would like to remove such restrictions from this index if possible..

Reply:
yes, inserts never go to logs right?

Reply:
for now.. you can assume this...